### ISSUE TITLE
ObjectEditor

### ISSUE THREAD

**Issue by glinsky007** [enthought/traitsui#1539]
When I right button mouse click to "Add New Item" (Default List of Objects editor) , nothing happens and I get the following Traceback.  If I have another button that adds an object to the list everything works as expected. 

```
Traceback (most recent call last):
  File "/Users/meglinsky/opt/anaconda3/lib/python3.8/site-packages/traitsui/qt4/table_model.py", line 246, in insertRows
    editor.callx(items.insert, row + i, editor.create_new_row())
  File "/Users/meglinsky/opt/anaconda3/lib/python3.8/site-packages/traitsui/qt4/table_editor.py", line 414, in create_new_row
    return self.ui.evaluate(
  File "/Users/meglinsky/opt/anaconda3/lib/python3.8/site-packages/traitsui/ui.py", line 696, in evaluate
    return eval(function, globals(), context)(*args, **kw_args)
TypeError: eval() arg 1 must be a string, bytes or code object
```

<img width="602" alt="Screen Shot 2021-03-13 at 8 00 42 PM" src="https://user-images.githubusercontent.com/5082847/111055824-f1841e00-8436-11eb-96b5-783559687758.png">

**Comment by rahulporuri** [comment]
Thanks for the report @glinsky007 . Again, can you report what version of python and traitsui you are using? Also, it'd be very useful if you can provide a minimal example that we can use to reproduce the issue.

**Comment by glinsky007** [comment]
Using Python 3.8.5, TraitsUI 7.1.1.  Works with Python 2.7.13, and 5.1.0-2.  Attached is some minimal code that repoduces the error.
[traits_error_demo.py.zip](https://github.com/enthought/traitsui/files/6141286/traits_error_demo.py.zip)

**Comment by glinsky007** [comment]
Added comments to issue.

**Comment by aaronayres35** [comment]
I’ve spent some time tracking this down and I believe this issue actually originates in Traits.  I have a “fix” currently but at this point it is a bit ugly.  

Recap of my findings (for future reference):
Following the traceback from the issue we look at https://github.com/enthought/traitsui/blob/2b01eaeda8db8c10d8fd21d3195c57d95225e4a4/traitsui/ui.py#L677-L689

 A print statement reveals that `function` is actually a tuple, hence the failure.   Taking one step further in the traceback we see that the above `function` comes from `factory.row_factory` here: 
 https://github.com/enthought/traitsui/blob/2b01eaeda8db8c10d8fd21d3195c57d95225e4a4/traitsui/qt4/table_editor.py#L424-L426
    this is the function called that creates the editor for this use case: https://github.com/enthought/traits/blob/2b62cd9b39bcb9e4998eeab0b32a4e3127b3df66/traits/editor_factories.py#L157-L192 

    and we see that `row_factory=_instance_handler_factory(item_handler)` 
See:
https://github.com/enthought/traits/blob/2b62cd9b39bcb9e4998eeab0b32a4e3127b3df66/traits/editor_factories.py#L143-L154
In this case, we have a `BaseInstance` (we started with a list of Instances of `ElemFraction`).  Thus, we are using the Instance’s `default_value`.  Instance traits can have their defaults specified in a couple of ways (ref: https://docs.enthought.com/traits/traits_user_manual/defining.html#instance).  
Looking at 
https://github.com/enthought/traits/blob/2b62cd9b39bcb9e4998eeab0b32a4e3127b3df66/traits/trait_types.py#L3353-L3356

We see that the default value can actually be an instance of `_InstanceArgs`  Thus, we need to do something like what is done here:
https://github.com/enthought/traits/blob/2b62cd9b39bcb9e4998eeab0b32a4e3127b3df66/traits/trait_types.py#L3395-L3399 

In this code:
https://github.com/enthought/traits/blob/2b62cd9b39bcb9e4998eeab0b32a4e3127b3df66/traits/editor_factories.py#L150-L151 
  Doing this: 
``` 
elif isinstance(handler, BaseInstance):
    result = handler.default_value
    if isinstance(result, tuple):
        return lambda : result[0](*result[1], **result[2])
    else:
        return result
```

Appears to resolve the issue for me locally.  Again, this is currently sort of ugly, as this is a first pass at finding /solving the problem

**Comment by glinsky007** [comment]
Has this been fixed?  When I am running Python 3.8.12 and TraitsUI 7.2.1, I am still getting the same error.
